### PR TITLE
Check if story and meeting exist

### PR DESCRIPTION
**Pull request by nickoferrall**
PR to resolve issues #4785 & #4708

### CODE REVIEW

**Comment by mattkrick** [comment]
any idea about the root cause here?

**Comment by nickoferrall** [comment]
For #4785, `Story` is nullable but we don't have a check to see if it exists (although the sidebar and control bar should prevent the user from advancing if there isn't as story). 

For #4708, I'm not too sure how an active meeting could be an empty array / not have an `id`

**Comment by mattkrick** [comment]
is a buggy optimistic updater possibly at fault? would love to reproduce!

**Comment by nickoferrall** [comment]
Interesting - could be, I'll take a look! Tomorrow morning I'll work on the GH issue but once that's ready or when I feel like switching things up, I'll dig into this!

**Comment by mattkrick** [comment]
@nickoferrall any update on this? would love to understand the root cause before merging the fix

**Comment by nickoferrall** [comment]
@mattkrick I haven't looked into this yet but I'll take a look now!

**Comment by nickoferrall** [comment]
I haven't been able to repro this yet but here's a few things that I've learnt so far:

1. As you mentioned, this could be a buggy optimistic updater issue, [like this one](https://github.com/ParabolInc/parabol/commit/af6c7ea717852cf402b996481896be51932d5ab5). I suspected that [this is where the problem lies](https://github.com/ParabolInc/parabol/blob/f092a703bd675aaaa5129d3e155c6d4cb51474e2/packages/client/mutations/UpdatePokerScopeMutation.ts#L152) but we can remove `nextEstimateStage.setLinkedRecord(story, 'story')` and story will still exist so it may not be the problem.

2. The Sentry events for the undefined `Story` happen in `PokerEstimateHeaderCardParabol` and not `PokerEstimateHeaderCardJira` because we check whether `Story` exists in the Jira header [here](https://github.com/ParabolInc/parabol/blob/22328ab58233bb34fb25a162c19ae29a5bc08f74/packages/client/components/PokerEstimateHeaderCardJira.tsx#L92). We've probably been showing the incorrect error message that Jira is down when actually it might be the optimistic updater at fault. 

3. This issue first appeared in release 6.1.0. In that release, we had two large PRs: [Freeze templates for meeting durarion](https://github.com/ParabolInc/parabol/pull/4695/files) and [Robust Meeting Queue Protocal](https://github.com/ParabolInc/parabol/pull/4606/files). I haven't noticed any changes in these PRs, or the other smaller ones, that could have caused this bug. The Freeze Templates PR touches the `UpdatePokerScopeMutation` with `dimensionRef`s but that shouldn't affect `Story`.

I'll keep looking into the optimistic updater in `UpdatePokerScopeMutation`, consider other mutations that might be at fault and see if I can repro the nullable Story.

File: packages/client/components/PokerEstimateHeaderCardJira.tsx [hunk 1]
```diff
@@ -83,8 +84,6 @@ const PokerEstimateHeaderCardJira = (props: Props) => {
   const {stage} = props
   const {serviceTaskId, story} = stage
   const [isExpanded, setIsExpanded] = useState(false)
-  const descriptionRef = useRef<HTMLDivElement>(null)
-  const maxHeight = Math.min(descriptionRef.current?.scrollHeight ?? 300, 300)
```
Comment (nickoferrall) [inline packages/client/components/PokerEstimateHeaderCardJira.tsx:87]: If the user clicks expand, I think we can show 100% of the height of the card. Calculating height like this didn't show the full error description

File: packages/client/components/PokerEstimateHeaderCardParabol.tsx [hunk 2]
```diff
@@ -87,31 +99,65 @@ const Content = styled('div')({
...
   const maxHeight = Math.min(descriptionRef.current?.scrollHeight ?? 300, 300)
   useEffect(
     () => () => {
       setIsExpanded(false)
     },
-    [taskId]
+    [story?.id]
   )
-  const {useTaskChild} = useTaskChildFocus(taskId)
+  const {useTaskChild} = useTaskChildFocus(story?.id || '')
```
Comment (nickoferrall) [inline packages/client/components/PokerEstimateHeaderCardParabol.tsx:144]: This feels ugly but if we check whether `story` exists beforehand and return the error card before `useEffect` or `useTaskChildFocus` are called, we won't be following the rules of hooks

Comment (mattkrick) [reply packages/client/components/PokerEstimateHeaderCardParabol.tsx:144]: i _really_ wish the rule of hooks didn't apply to returning `null`.
the least ugly pattern i've found is `const storyId = story?.id ?? ''` at the top of the component

Comment (nickoferrall) [reply packages/client/components/PokerEstimateHeaderCardParabol.tsx]: Makes sense! I've moved it to the top which feels slightly less ugly

File: packages/client/components/PokerEstimateHeaderCardParabol.tsx [hunk 3]
```diff
@@ -87,17 +98,37 @@ const Content = styled('div')({
...
 }
 
-type Story = Required<NonNullable<NonNullable<PokerEstimateHeaderCardParabol_stage>['story']>>
-
 const PokerEstimateHeaderCardParabol = (props: Props) => {
   const {stage} = props
-  const {story} = stage
-  const {content, id: taskId, teamId} = story as Story
-  const integration = story!.integration
+  const story = stage.story as Extract<typeof stage['story'], {__typename: 'Task'}> | null
```
Comment (mattkrick) [inline packages/client/components/PokerEstimateHeaderCardParabol.tsx:129]: @nickoferrall i know we weren't happy with how we were doing this, so i found a neat little hack. if you request the `__typename` field explicitly in the fragment, then relay gives you a discriminated union. you can then `Extract` the type you want, which means nullability for the story fields will be preserved.

File: packages/client/hooks/useSnacksForNewMeetings.ts [hunk 4]
```diff
@@ -32,7 +32,7 @@ const useSnacksForNewMeetings = (meetingsRef: any) => {
     const fiveSecsAgo = new Date(Date.now() - ms('5s'))
     const sortedMeetings = meetings
       .filter((meeting) => {
-        if (dismissedMeetingIds.has(meeting.id)) return false
+        if (dismissedMeetingIds.has(meeting?.id)) return false
```
Comment (mattkrick) [inline packages/client/hooks/useSnacksForNewMeetings.ts:35]: the next line will throw if meeting is undefined

File: packages/client/components/PokerEstimateHeaderCardParabol.tsx (hunk 3 above)
Comment (nickoferrall) [reply packages/client/components/PokerEstimateHeaderCardParabol.tsx:129]: Nice! That's good to know

File: packages/client/hooks/useSnacksForNewMeetings.ts (hunk 4 above)
Comment (nickoferrall) [reply packages/client/hooks/useSnacksForNewMeetings.ts]: Updated!

**Comment by mattkrick** [comment]
Screenshot from 2021-03-31 08-20-46

when collapsed, i see the expand button as well as scrollbars.

I'd suggest:
- hide the scroll bars when collapsed
- show the scroll bars when expanded
- set a max height when expanded so the dimension is still above the fold

**Comment by mattkrick** [comment]
great work! everything looks like it's working as intended!

**Pull request merged**